omnigibson/utils/bddl_utils.py

```diff
@@ -89,9 +127,810 @@ def process_single_condition(condition):
+
+    def _maximum_bipartite_matching(self, filtered_object_scope, condition_type):
+        """
+        Matches objects from @filtered_object_scope to specific room instances it can be
+        sampled from
+
+        Args:
+            filtered_object_scope (dict): Filtered object scope
+            condition_type (str): What type of condition to sample, e.g., "initial"
+
+        Returns:
+            None or str: If successful, returns None. Otherwise, returns an error message
+        """
+        # For each room instance, perform maximum bipartite matching between object instance in scope to simulator objects
+        # Left nodes: a list of object instance in scope
+        # Right nodes: a list of simulator objects
+        # Edges: if the simulator object can support the sampling requirement of ths object instance
+        for room_type in filtered_object_scope:
+            # The same room instances will be shared across all scene obj in a given room type
+            some_obj = list(filtered_object_scope[room_type].keys())[0]
+            room_insts = list(filtered_object_scope[room_type][some_obj].keys())
+            success = False
+            # Loop through each room instance
+            for room_inst in room_insts:
+                graph = nx.Graph()
+                # For this given room instance, gether mapping from obj instance to a list of simulator obj
+                obj_inst_to_obj_per_room_inst = {}
+                for obj_inst in filtered_object_scope[room_type]:
+                    obj_inst_to_obj_per_room_inst[obj_inst] = filtered_object_scope[room_type][obj_inst][room_inst]
+                top_nodes = []
+                log_msg = "MBM for room instance [{}]".format(room_inst)
+                log.debug((log_msg))
+                for obj_inst in obj_inst_to_obj_per_room_inst:
+                    for obj in obj_inst_to_obj_per_room_inst[obj_inst]:
+                        # Create an edge between obj instance and each of the simulator obj that supports sampling
+                        graph.add_edge(obj_inst, obj)
+                        log_msg = "Adding edge: {} <-> {}".format(obj_inst, obj.name)
+                        log.debug((log_msg))
+                        top_nodes.append(obj_inst)
+                # Need to provide top_nodes that contain all nodes in one bipartite node set
+                # The matches will have two items for each match (e.g. A -> B, B -> A)
+                matches = nx.bipartite.maximum_matching(graph, top_nodes=top_nodes)
+                if len(matches) == 2 * len(obj_inst_to_obj_per_room_inst):
+                    log.debug(("Object scope finalized:"))
+                    for obj_inst, obj in matches.items():
+                        if obj_inst in obj_inst_to_obj_per_room_inst:
+                            self._object_scope[obj_inst] = BDDLEntity(object_scope=obj_inst, entity=obj)
```

Comment:
we should do `self._object_scope[obj_inst].bddl_object_scope = obj_inst` here if needed. We can discuss in person!

Reply:
outdated